### PR TITLE
Initial version of the hazard mechanism proposed in #133

### LINKED ISSUE #133: Would be nice to support hazard pointer like claims on heap objects
We have a notion of claims of heap objects, but they're fairly heavy, requiring a full compartment call to either acquire or drop.  We'd like a lighter-weight option for a small number of objects per thread: hazard pointers understood by the heap.

Hazard pointers originate in the world of lock/wait-free data structures, where pointers have a four-stage lifecycle: unpublished, public, withdrawn, and freed.  Hazard pointers mediate the transition from withdrawn to freed, deferring it until there are no more hazardous, local copies of once-public pointers.  Because everyone involved is presumed cooperative, hazard lists need only be scanned once: a request to free memory must follow that pointer being withdrawn from the shared state, and so it cannot become newly hazardous.

The story for potentially uncooperative threads is slightly different, as it may be impossible to ensure that a pointer is completely withdrawn from the shared state (and is not cached somewhere without being declared hazardous) when it comes time to `free()` it.  This is, after all, why we have sweeping revocation!  As such, our story needs to be a little different.  In particular, we need a ratchet mechanism to advance an object through these states:
1. allocated and hazard-able (and possibly hazardous)
2. allocated but no longer hazard-able, with N hazards registered
3. ...
4. allocated but no longer hazard-able, with 1 hazard registered
5. No longer hazardous, and so free (well, quarantined)

Additionally, hazards are intended to be ephemeral.  As such, we do not expect them to survive cross-compartment calls.  (And anyway, a callee cannot assume that its caller has declared any hazards appropriately, and so must do so itself; if we ensured that hazards persisted across calls, there would be redundant registrations.)

Here is a design sketch.  This is going to need to be a fair bit of carefully crafted assembler but I believe it's entirely possible.

There is a Hazard Registry maintained on behalf of the allocator by a privileged _library_.  A Hazard Registry Entry (or just "Hazard") is an optional pair of a capability to a (sub)object and a taint flag.  The `None` branch of the option is indicated by a `null` cap.  Hazard entries are, we imagine, stored in static sealed objects, with handle type `SealedHeapHazard`, each of which has room for some number of hazards.  To avoid linked-list traversal, we imagine that these are collected in a dedicated section, forming a "linker set".

The privileged library exposes a single call to clients (and more to the allocator itself):
`void *heap_hazard_register(SealedHeapHazards *hh, size_t ix, void *p)`, with interrupts deferred,
  - unseals `hh`,
  - releases the existing hazard in the `ix`-th entry in `hh`, if any, and
  - scans all existing hazards, looking for a tainted entry that precludes registering `p` (see below), and
  - if `p` is tagged, installs `p` into that entry and and returns `p`.
If `p` is untagged or has a tainted hazard, this call returns `null`.  The registry entries are _write-only_ from the clients' perspective: it is not possible to retrieve a hazard pointer.
 
Releasing an untainted hazard is as simple as replacing it with `null`.  Releasing a tainted hazard triggers a full, interrupt-enabled cross-compartment call to the allocator to advance the effort to free the associated object.  Importantly, the hazard remains tainted while this call is in progress so that the taint is seen by other calls to `heap_hazard_register`.  Specifically, this advance involves scanning the set of hazards, and either
- finding no relevant hazards; proceed to quarantine the object
- propagating taint: taint some other, relevant hazard, upgrading the pointer in that entry to be the full object
before replacing the hazard with `null`.

Upon discovering that there are no more claims to an object, the allocator will effectively register this object as a _tainted_ hazard and then immediately release that hazard (that is, scan the hazard table looking for another entry to taint).  One could imagine a dedicated `SealedHeapHazard` chunk of the registry specifically for `free()`'s use, but it might make more sense to special case this primordial tainting.

Tainted hazards are always carrying pointers full heap objects, while untainted ones may have sub-object pointers; this simplifies the scan above without unduly challenging the initial sweep for hazards.

Commentary welcome.

### PR DESCRIPTION
**Pull request by davidchisnall**
This provides a lightweight claims mechanism where each thread has two slots that can be used to store pointers to objects that have been temporarily claimed.  These are automatically cleared on cross-compartment call.

A Switcher function allows code to retrieve a (local) pointer to the current thread's space.  This is local so that it cannot be shared between threads (doing so would allow one thread to clear another thread's hazard slots while that thread entered another compartment). It is store-only to provide some defence in depth.

The hazard slots are synchronised with the allocator via an epoch mechanism.  The allocator increments the epoch when it starts scanning the list and again at the end.  If the value of the epoch counter changes during insertion or is odd then, the thread trying to insert needs to retry.  There is currently an unconditional futex wake on the allocator's free path to allow callers to sleep on the futex.  Given how rare contention is expected to be, it's possibly better to just sleep for one tick when this happens and avoid the notify.  This is a simple change to make.

This structure is not intended to be manipulated directly.  The `heap_claim_fast` library function provides a user-friendly wrapper around it.

Along the way, fixes a bug where quotas could be amplified.

Fixes #133

### CODE REVIEW

**Comment by davidchisnall** [comment]
Some optimisation later:

```
	Original	New		Overhead
32	110578928	123680994	11.85%
64	58579947	65044143	11.03%
128	30624678	34142932	11.49%
256	16749468	18369632	9.67%
512	9623171		10526866	9.39%
1024	6562620		6945096		5.83%
2048	4268676		4531561		6.16%
4096	3342736		3479603		4.09%
8192	2618123		2667895		1.90%
16384	2184942		2209965		1.15%
32768	1955510		1966652		0.57%
65536	2554743		1910479		-25.22%
131072	2616802		3092058		18.16%
```

This version uses a futex_wait on the waiting side, but with a one-tick timeout and with priority boosting, so it should always kick the freeing thread out of free block in one tick.  It doesn't require the freeing side to do anything (we rely on the one-tick timeout to exit).  

With this speed, we can manage around 3 MiB/s allocation rates for 1024 byte allocations and over 300 KiB/s for 64-byte allocations at 20 MHz, which seems plenty.  In particular, it should be sufficient for 10 Mb/s networking with one malloc per packet.

File: sdk/core/allocator/main.cc [hunk 1]
```diff
@@ -782,6 +797,13 @@ namespace
 		  *capability, *chunk, bodySize, isPrecise, reallyFree);
 	}
 
+	bool timeout_is_valid(Timeout *timeout)
+	{
+		return !heap_address_is_valid(timeout) &&
```
Comment (rmn30) [inline sdk/core/allocator/main.cc:802]: This is worthy of a comment, at least and should be documented on the APIs that take a `Timeout`. By the way I observe that `heap_address_is_valid` returns true for pointers into the hazardQuarantine but I think that is fine as they should never escape the allocator.

File: sdk/core/loader/boot.cc [hunk 2]
```diff
@@ -758,6 +758,16 @@ namespace
...
+		Capability<void *> hazardPointers =
+		  build<void *,
+		        Root::Type::RWGlobal,
+		        PermissionSet{Permission::Store,
+		                      Permission::LoadStoreCapability}>(
+		    LA_ABS(__export_mem_hazard_pointers),
+		    LA_ABS(__export_mem_hazard_pointers_end) -
+		      LA_ABS(__export_mem_hazard_pointers));
+		// Space per thread for hazard pointers.
+		static constexpr size_t HazardPointerSpace = 2 * sizeof(void *);
```
Comment (rmn30) [inline sdk/core/loader/boot.cc:770]: Could we have a symbolic name for the number of hazard pointers per thread? Could it be configurable?

Comment (davidchisnall) [reply sdk/core/loader/boot.cc:770]: Do you imagine ever supporting more than 2?  I think 2 is both the minimum number to be useful and the maximum number to be performant.

File: sdk/core/allocator/main.cc (hunk 1 above)
Comment (davidchisnall) [reply sdk/core/allocator/main.cc:802]: Agreed.  I think that should go in timeout.h in the generic coverage of timeouts.  In a follow-on, I am also going to add a check-timeout helper like this, but I want to see with respect to code size whether it makes more sense in the helpers library or as a static inline.

File: sdk/core/loader/boot.cc (hunk 2 above)
Comment (rmn30) [reply sdk/core/loader/boot.cc:770]: Perhaps not, but it would make the code more self-documenting rather than a magic constant. Bonus points if the same name could be used in the linker script.

Comment (davidchisnall) [reply sdk/core/loader/boot.cc:770]: Yup, I agree it should be a symbolic constant, just not sure what scope it needs.  If you agree that we won't change 2, then it can go here.  

Unfortunately, the linker script is generated from lua and we don't have a mechanism for sharing constants between the build system and the code.  It would be nice to add one.

File: sdk/core/allocator/main.cc (hunk 1 above)
Comment (davidchisnall) [reply sdk/core/allocator/main.cc:802]: Comment added in timeout.h

File: sdk/core/loader/boot.cc (hunk 2 above)
Comment (davidchisnall) [reply sdk/core/loader/boot.cc]: Symbolic constant added.

File: sdk/boards/ibex-arty-a7-100.json [hunk 3]
```diff
@@ -23,6 +23,14 @@
         "gpio_led0" : {
             "start": 0x8f00f000,
             "length": 0x800
+        },
+        "dmb_cfg": {
```
Comment (rmn30) [inline sdk/boards/ibex-arty-a7-100.json:27]: What's this?

File: sdk/core/allocator/alloc.h [hunk 4]
```diff
@@ -891,6 +902,46 @@ class MState
...
+	 */
+	bool hazard_quarantine_is_empty()
+	{
+		return hazardQuarantineSize == 0;
+	}
+
+	/**
+	 * Returns true of `ptr` is already in the quarantine list.
+	 */
+	bool hazard_quarantine_contains(Capability<void> ptr)
```
Comment (rmn30) [inline sdk/core/allocator/alloc.h:921]: There don't seem to be any calls to this function? I was a bit worried about the `is_subset_of`.

File: sdk/boards/ibex-arty-a7-100.json [hunk 5]
```diff
@@ -23,6 +23,14 @@
...
+            "start": 0x8f0f0000,
+            "length": 0x1000
+        },
+        "dmb": {
+            "start": 0x90000000,
+            "length": 0x100000
```
Comment (nwf-msr) [inline sdk/boards/ibex-arty-a7-100.json:33]: This hunk looks unrelated; am I missing something?

File: sdk/core/allocator/alloc.h [hunk 6]
```diff
@@ -1160,26 +1250,135 @@ class MState
...
+			 * that they are waiting for us to walk the epoch list will block
+			 * until we're done.
+			 *
+			 * It must also happen before zeroing because the allocator is
+			 * running with interrupts enabled.  Were we to zero and then paint,
+			 * there would be a window in which preemption could allow a store
+			 * through a copy of the user capability (or its progeny) that undid
+			 * our work of zeroing!
+			 */
+			revoker.shadow_paint_range<true>(mem.address(), chunk.cell_next());
```
Comment (nwf-msr) [inline sdk/core/allocator/alloc.h:1380]: Perhaps a comment of `guard` going out of scope after this line, just so there's _something_ written that makes it look like this can't be CSE'd with the then branch above.

File: sdk/core/allocator/main.cc [hunk 7]
```diff
@@ -732,12 +735,22 @@ namespace
...
+				if (ret == 0)
+				{
+					owner.quota += chunkSize;
+				}
+				return ret;
 			}
-			chunk.ownerID = 0;
+			// We've removed the owner, so refund the quota immediately, the
+			// free won't happen until the last claim goes away, but this is no
+			// longer the owner's responsibility.
```
Comment (rmn30) [inline sdk/core/allocator/main.cc:752]: A bit unfortunate that this means there is memory that is not free (could remain allocated indefinitely) but isn't attributed to a quota. With the existing claim mechanism does it get charged to the claimer's quota?

File: sdk/core/allocator/alloc.h [hunk 8]
```diff
@@ -891,6 +902,46 @@ class MState
 	size_t heapFreeSize;
 	size_t heapQuarantineSize;
 
+	/**
+	 * The number of entries currently in the `hazardQuarantine` array.
+	 */
+	size_t hazardQuarantineSize = 0;
```
Comment (nwf-msr) [inline sdk/core/allocator/alloc.h:908]:
```suggestion
	size_t hazardQuarantineOccupancy = 0;
```

since you use `hazardQuarantineSize` for the actual size of the reserved region, below?

File: sdk/lib/compartment_helpers/claim_fast.cc [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+		values--;
+	}
+	if (!heap_address_is_valid(ptr2))
+	{
+		ptr2 = nullptr;
+		values--;
+	}
+	// If neither pointer refers to heap memory, set without synchronizing.  It
+	// doesn't matter if the revoker sees these or not, they cannot extend the
+	// lifetime of a heap object.
```
Comment (rmn30) [inline sdk/lib/compartment_helpers/claim_fast.cc:31]: Is this comment a bit stale? I don't understand it and I'm not sure it describes the code.

File: sdk/include/stdlib.h [hunk 10]
```diff
@@ -145,6 +145,27 @@ void *__cheri_compartment("alloc")
...
+ * A null pointer can be used as a not-present value.  This function will
+ * treat operations on null pointers as unconditionally successful.  It returns
+ * `-ETIMEDOUT` if it failed to claim before the timeout expired or `EINVAL` if
+ * one or more of the arguments is neither null nor a valid pointer at the end.
+ *
+ * In the case of failure, neither pointer will have been claimed.
+ *
+ * This function is provided by the compartment_helpers library, which must be
+ * linked for it to be available.
+ */
```
Comment (rmn30) [inline sdk/include/stdlib.h:164]: Is the idiomatic way to release a claim to call with `nullptr`?

**Review by nwf-msr — COMMENTED** [review]
Nits, but generally LGTM

**Review by rmn30 — APPROVED** [review]

File: sdk/boards/ibex-arty-a7-100.json (hunk 5 above)
Comment (davidchisnall) [reply sdk/boards/ibex-arty-a7-100.json:33]: Yup, that was accidentally added.

File: sdk/core/allocator/alloc.h (hunk 6 above)
Comment (davidchisnall) [reply sdk/core/allocator/alloc.h:1380]: Does the comment above not say this?  It starts with 'Paint the shadow bitmap with the hazard epoch odd'.  I've added a comment to the guard.

File: sdk/boards/ibex-arty-a7-100.json (hunk 3 above)
Comment (davidchisnall) [reply sdk/boards/ibex-arty-a7-100.json:27]: Accidental things from the network stack bits.

File: sdk/lib/compartment_helpers/claim_fast.cc (hunk 9 above)
Comment (davidchisnall) [reply sdk/lib/compartment_helpers/claim_fast.cc:31]: No, this refers directly to the code immediately below it.  Above, we've checked whether either pointer refers to heap memory.  Below, if they don't, we null out the hazard slots and return immediately.

File: sdk/include/stdlib.h (hunk 10 above)
Comment (davidchisnall) [reply sdk/include/stdlib.h:164]: Yup.  Or simply return from a cross-compartment call (or call another cross-compartment call).  For example, if you're yielding in a loop, you implicitly release them before each loop iteration, so you don't need the explicit call on retry.

File: sdk/core/allocator/alloc.h (hunk 6 above)
Comment (nwf-msr) [reply sdk/core/allocator/alloc.h:1368]: Sorry, I'd meant a comment _after_ the call to `shadow_paint_range` so that someone like future-me, reading upwards from the ends of the then and `else` scopes wouldn't forget that `guard` going out of scope was important.
